### PR TITLE
[4.0] use standard SQL - fix errors on postgresql

### PR DESCRIPTION
**Pull request by alikon**
Pull Request for Issue 
use standard SQL

### Summary of Changes
rewrite query in standard SQL instead of mysql dialect
- correct JOIN   for syntax error at or near "WHERE" 
- ERROR: CASE types character varying and numeric cannot be matched

### Testing Instructions

Install latest 4.0-dev and populate data with sample testing plugin
go on frontpage

### Expected result

see articles even in postgresql 

### Actual result
errors with postgresql

### CODE REVIEW

**Comment by richard67** [comment]
Testing instructions say “install latest staging” but I guess you mean “install latest 4.0-dev” since this PR is for 4.0-dev. If I am right, please correct.

**Comment by alikon** [comment]
yes latest 4.0

**Comment by richard67** [comment]
Will test later in the afternoon CEST time.

File: plugins/content/pagenavigation/pagenavigation.php [hunk 1]
```diff
@@ -135,11 +135,11 @@ public function onContentBeforeDisplay($context, &$row, &$params, $page = 0)
 
 			$case_when = ' CASE WHEN ' . $query->charLength('a.alias', '!=', '0')
 				. ' THEN ' . $query->concatenate(array($query->castAsChar('a.id'), 'a.alias'), ':')
-				. ' ELSE a.id END AS slug';
+				. ' ELSE '. $query->castAsChar('a.id') . ' END AS slug';
```
Comment (richard67) [inline plugins/content/pagenavigation/pagenavigation.php]:
```suggestion
				. ' ELSE ' . $query->castAsChar('a.id') . ' END AS slug';
```

File: plugins/content/pagenavigation/pagenavigation.php [hunk 2]
```diff
@@ -135,11 +135,11 @@ public function onContentBeforeDisplay($context, &$row, &$params, $page = 0)
...
 
 			$case_when1 = ' CASE WHEN ' . $query->charLength('cc.alias', '!=', '0')
 				. ' THEN ' . $query->concatenate(array($query->castAsChar('cc.id'), 'cc.alias'), ':')
-				. ' ELSE cc.id END AS catslug';
+				. ' ELSE '. $query->castAsChar('cc.id') . ' END AS catslug';
```
Comment (richard67) [inline plugins/content/pagenavigation/pagenavigation.php]:
```suggestion
				. ' ELSE ' . $query->castAsChar('cc.id') . ' END AS catslug';
```

**Review by richard67 — CHANGES_REQUESTED** [review]
Fix PHPCS errors reported by drone.

**Comment by richard67** [comment]
@alikon If you find time, please fix the drone PCPCS errors.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25724">issues.joomla.org/tracker/joomla-cms/25724</a>.</sub>

**Comment by alikon** [comment]
@richard67 cs fixed..... when you have time...

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 356f56fb3934d789c2433a730b2613be3e3e853f<br /><br />There are still errors with testing sample data on PostgreSQL, but they are not related to this PR, i.e. the errors are there with and without this PR:

1. On MySQL, the article with title "Joomla! Testing" as 4 tags assigned, tags "2", "3", "4" and "5", but on PostgreSQL the same article has only 1 assigned, which is tag "2".
2. When editing any article, a red popup alert "Error" is shown, telling only "An error has occurred.".
3. When trying to save a new article, an SQL errors about null value for column with not null constraint is shown.

As I said, it seems not to be related to this PR.

I will make new issues for that, except @alikon tells me before that he wants to fix the one or other thing here in this PR, too.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25724">issues.joomla.org/tracker/joomla-cms/25724</a>.</sub>

**Comment by alikon** [comment]
@richard67  thanks for testing
it should be always better to have 1 pr for 1 issue as a main rule
so i will be happy if free time permits me, to make a pr for any issue postgresql related (even not) that you'll submit  (you know my mantra is to close issues as many as possible :smile: )

**Comment by richard67** [comment]
@alikon Could you check issue #25747 ? ~~It seems to be related to SQL in testing sample data, too, so maybe this PR here would be the right place so fix that.~~ Edit: No, I think meanwhile error is somewhere else, maybe in tags component.

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**